File: code/python/309.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        dp_i_0 = 0
        dp_i_1 = float('-inf')
        dp_i_0_1 = 0
        dp_i_1_1 = float('-inf')
        for price in prices:
            temp_0 = dp_i_0
            temp_1 = dp_i_1
            dp_i_0 = max(dp_i_0,dp_i_1+price)
            dp_i_1 = max(dp_i_1,dp_i_0_1-price)
            dp_i_0_1 = temp_0
            dp_i_1_1 = temp_1
            #print(dp_i_0,dp_i_1)
        return dp_i_0

File: code/java/309.java
class Solution {
    public int maxProfit(int[] prices) {
        int dp_i_0 = 0;
        int dp_i_1 = Integer.MIN_VALUE;
        int dp_i_0_0 = 0;
        int temp = 0;
        for(int i=0;i<prices.length;i++){
            dp_i_0_0 =temp;
            temp = dp_i_0;
            dp_i_0 = Math.max(dp_i_0,dp_i_1+prices[i]);
            dp_i_1 = Math.max(dp_i_1,dp_i_0_0-prices[i]);
        }
        return dp_i_0;
    }
}